### ISSUE TITLE
Parse network stream

### ISSUE THREAD

**Issue by nickbabcock** [nickbabcock/boxcars#6]
This would be 90% of the file, and it's a shame that my enthusiasm for
implementing this section waned. When the developers of the game say the
section isn't easy to parse, the major rocket league libraries dedicate half of
their code to parsing the section, and the with each patch everything breaks,
it's an incredible feat for anyone to retain enthusiasm. Way to go maintainers!

We already extracted most of the interesting bits like player stats and goals
contained in the header, so it's not a tremendous loss if we can't parse the
network data. If we were able to parse the network data, it would allow us to
run benchmark against other implementations.

**Comment by tfausak** [comment]
Hi! I discovered this repository today. I think writing a parser in Rust makes a lot of sense! I don't know Rust that well, but I'm pretty familiar with Rocket League's replay file format. I would be happy to give you tips and translate my obtuse Haskell code into English and/or Rust. 

The replay file format doesn't change all that often. In fact, it's surprising how little it's changed considering they've added big features like rumble mode and drop shot! At any rate, it doesn't normally take that long to figure out the new changes. The worst part is the lack of notice or help from the developers.

It's true that most of the interesting data lives in the replay headers these days. However, I am aware of many people using replays for machine learning. That requires the actual network stream data. 

If the benchmarks in your README are to be believed, Boxcars has a good chance at being *much* faster than Rattletrap. We've come a long way from parsing a replay with Python in 30 seconds, and it'd be great to go even farther!

**Comment by nickbabcock** [comment]
> I'm pretty familiar with Rocket League's replay file format

You and I know that is both an understatement 😉 

> I'm pretty familiar with Rocket League's replay file format

Most of the code is using the [parser combinator library, nom](https://github.com/Geal/nom), so it doesn't really look like typical rust. This contributes to part of my mixed opinions about that library. Since I'm experienced with writing parsers by hand, I thought about rewriting it -- in part to achieve better error handling.

> The worst part is the lack of notice or help from the developers.

Right, so it sounds like a proper design would return a Success even if network data failed so that it 
guards against breaking changes in the network data. 

> However, I am aware of many people using replays for machine learning. That requires the actual network stream data.

Wow! That's really cool. I'm already impressed by the heatmaps people make from the network data.

> If the benchmarks in your README are to be believed

You better believe them 😈 

Buuuut that might be because when you exclude the crc check and network data there isn't all that much to parse 😄 

I'll see if I can't adapt some of your network code to get a better idea of what's involved.

I can't thank you enough for your Octane library, which I heavily leaned on for this.

**Comment by nickbabcock** [comment]
- I replaced the previous parsing logic and found a 2x decrease in time it takes to parse the header and footer. All contained within [parsing.rs](https://github.com/nickbabcock/boxcars/blob/5ee8cc28080a01808b6e1ca1224de6507b1f53e1/src/parsing.rs). This code should be more approachable to a Rust beginner
- Added an option to only run the crc check on a section if parsing failed (as the crc check is 100x the cost of the actual parsing)
- Parsing the header and footer takes 30 microseconds, so I'm confident that parsing the network data won't exceed 1ms.

I was diving into the network code using your and jjbott's repos. It is a little overwhelming, you don't happen to have a forum / blog post that breaks down kinda what to expect for the network data 😅  (I keep losing my mental model as I jump between all the files)?

**Comment by tfausak** [comment]
I haven’t written a blog post because the audience would be tiny. Most of the complexity is in the class property map, which is a mess. After that, each attribute can be taken more or less in isolation. Rattletrap has a bunch of replays, each of which exposes (usually) one new concept. So you may be able to work your way through the replays starting with the one called “no frames”. I’m on my phone right now, but I’ll take a closer look and come up with a better suggestion later. 

Parsing the network data in 1ms would be *incredible*. Rattletrap is fast, but parsing is on the order of 1MB/s if I remember correctly.

**Comment by tfausak** [comment]
I don't have any prose that explains the stream or frames, but this ancient document covers most of the other stuff at a high level: https://gitlab.com/taylorfausak/octane/blob/5c6334b/README.md

A decent way to understand the structure of the replay files is to follow the module documentation. There's not a lot of descriptive text in the it, but the hosted documentation might be easier to follow than clicking around in the code. It also has links for viewing the source code, so you can jump down to that level if you want to. It makes sense to start at the top with the replay: https://hackage.haskell.org/package/rattletrap-3.1.2/docs/Rattletrap-Replay.html

There is a page on the Rocket League Replays wiki that lists a bunch (but not all!) of the replay parses. It also includes links to discussions on Reddit and the Psyonix forums: https://github.com/rocket-league-replays/rocket-league-replays/wiki/Rocket-League-Replay-Parsers/13e3ae1273acc531ccc46d08f805c08e7a700d3d

The thread on the Psyonix forum is where all of this started. It remains a good resource, even though it's kind of hard to follow and contains some things that may technically be true but don't really help you understand anything. Nevertheless, it's probably worth a read: https://www.psyonix.com/forum/viewtopic.php?f=33&t=13656

Looking at the history for replay files in Rattletrap, you can find a path that might make sense. It starts with only two replays: one with no frames at all (which isn't possible anymore) and one that's as short and small as possible. It goes on to add attributes more or less one at a time by introducing replays that are slightly more complex. This isn't an easy resource to navigate, but there's lots of useful information encoded here: https://github.com/tfausak/rattletrap/commits/65d1980?after=65d19801b9bdca3d2ec95514185f24f68ce1ce60+34&path[]=test&path[]=replays

To sum up, if you're looking for a pile of code to glean information from, starting with the most basic version of Rattletrap wouldn't be a bad idea. You can browse the tree here: https://github.com/tfausak/rattletrap/commit/41f71fd46dfffdd650ad0bd4c9a50f75df590308

**Comment by nickbabcock** [comment]
Thank you for all those resources! I'll see if I can't whip up a strong foundation from what you provided. I'm playing around in the [network branch](https://github.com/nickbabcock/boxcars/tree/network) and ping back with progress

**Comment by tfausak** [comment]
Looks like you've made quite a bit of progress. Nice work! 

Out of curiosity, can you benchmark Boxcars against Rattletrap? I fully expect Boxcars to blow Rattletrap out of the water. I want to see just how much faster it is :)

**Comment by nickbabcock** [comment]
Wow after spending wayyyy too much time on this, I have an update. The network code that I just squashed into master parses 185 out of a 320 replay corpus (so there is still a lot of work to do), _but_ I figured I'd give you my thoughts regarding performance.

Remember when I said:

> I'm confident that parsing the network data won't exceed 1ms

I should have realized that if calculating the crc takes over 1ms, there is no way parsing the body will be under 1ms.

Current performance numbers:

- Parse header: 30us
- Parse header + body: 10-15ms
- Parse header + body + crc: 12-17ms
- Parse header + body + crc + output json: 30-40ms

Bottleneck is doing bitshifts when decoding rigid body attributes.

Recommendations for you:

- As far as I know, Rattletrap and RocketLeagueReplayParser use a dictionary / map of (strings to x) or (x to strings). There should be no strings in the frame decoding loop (outside of the string attribute). Calculating the hash of a string and string equality is extremely expensive when compared to integer hashing and comparison. What I've done is take all the class names you've defined in `Data` and map them to the object's class name index from the replay file, so that resolving attribute / new actors decoding is only an integer lookup away.
- No logarithm or non-power-of-two calculations inside frame decoding loop
- Other small optimizations that may add up, but not worth listing here

The big one here are strings, I suspect you can easily get a 10x performance boost by eschewing all strings in the frame decoding loop. Only resolve strings on errors, so you don't pay the price unless you need the context.

More thoughts:
- I created my own bitstream implementation (called bitter). None of the current Rust bit libraries were optimized for RL replays. I have mixed feelings about it. It's fast -- I found it best to process in 64 bit chunks as I could decode several fields in just one chunk (64 bit chunks makes bitter partial to 64bit platforms). The API contains both a safe and unsafe (functions that can [`panic`](https://doc.rust-lang.org/std/macro.panic.html)). The functions that can panic are a smidge faster, so when I add in more robust error handling I expect performance to drop somewhere in the area of 5%. I typically like to rely on libraries created by people smarter than me!
- How you all figure out class hierarchies, new attributes, etc is beyond me. I initially tried constructing the hierarchy as jjbott did (by computing it from the hierarchy stored in the replay), but grew frustrated when I encountered nonsense. Sooo, I just used your method, which ignores the hierarchy in the replay. I don't think I could derive the hierarchy you created from a blank slate, and it would be discouraging when new classes come out and break existing code. This makes me worried about maintenance, as my current methodology when I encounter a replay I can't parse is to run it through yours! 
- Still tons of work needed (error handling, code cleanup, etc)

**Comment by tfausak** [comment]
Ah, looking things up by ID rather than name makes a lot of sense! Honestly I haven't put that much effort into optimizing Rattletrap because it's been Fast Enough ™️ for me. Obviously faster is better, but I assumed that I wouldn't be able to wring out much more performance because of Haskell's GC.

I'm not surprised that you had to roll your own bit-level parser. One does happen to exist for Haskell, but it's not nearly as well polished as the byte-level parser. Most things are byte aligned 🤷‍♂️

It's been a long time since I actually looked at how I come up with the class/object/attribute hierarchy. It could be complete garbage. I think some other parsers do string manipulation to determine attribute types, but that felt brittle to me. That being said, coming up with the mapping between attribute names and the expected type is pretty much trial and error.

**Comment by tfausak** [comment]
I wrote a micro-benchmark comparing CRC32 implementations in Haskell and Rust: https://gist.github.com/tfausak/e80def1ea2494902bd4a83dc928bf577

Rust is *way* faster. Even if I optimized the heck out of Rattletrap, I don't think I could get it to be as fast as Boxcars.

**Comment by tfausak** [comment]
Oops, I spoke too soon. I ran the Haskell CRC benchmark without full optimizations (`-O2`). Once I turned those on, Haskell and Rust finished in about the same amount of time. So it may be possible for Rattletrap to be as fast as Boxcars 😄

**Comment by nickbabcock** [comment]
> So it may be possible for Rattletrap to be as fast as Boxcars 😄

Bring it on 😈 

I believe you hashed it this out in another thread but the table used in the crc is from [CRCTable_DEPRECATED](https://docs.unrealengine.com/latest/INT/API/Runtime/Core/Misc/FCrc/CRCTable_DEPRECATED/index.html)

The table can also be generated with:

**Comment by nickbabcock** [comment]
Oof accidentally hit Shift + Enter and my comment was posted prematurely. 

> So it may be possible for Rattletrap to be as fast as Boxcars 😄

Bring it on 😈 

Not to derail too far, but on the topic of crc calculations. I believe you hashed it this out in another thread but the table used in the crc is from [CRCTable_DEPRECATED](https://docs.unrealengine.com/latest/INT/API/Runtime/Core/Misc/FCrc/CRCTable_DEPRECATED/index.html)

The table can also be generated by pycrc v0.9.1, https://pycrc.org using the configuration (as you know already):
 - Width         = 32
 - Poly          = 0x04c11db7
 - XorIn         = 0x10340dfe
 - ReflectIn     = False
 - XorOut        = 0xffffffff
 - ReflectOut    = False

Crc calculation can be faster when one employs "Slice by 4/8" techniques (highlighted in the link you posted).

Running pycrc.py with `--slice-by 8` returned an error:

```
pycrc: warning: disabling slice-by for non-reflected algorithm
```

I only bring this up as the unreal docs list [CRCTablesSB8_DEPRECATED](https://docs.unrealengine.com/latest/INT/API/Runtime/Core/Misc/FCrc/CRCTablesSB8_DEPRECATED/index.html), but doesn't actually list the table! So pycrc can't compute it, but the unreal code thinks it's doable. Makes me confused 😕

**Comment by tfausak** [comment]
I haven't spent much time thinking about the CRC because it's not the bottleneck for Rattletrap. The [Fast CRC32 page](http://create.stephan-brumme.com/crc32/) does show higher speeds for methods that slice by 4, 8, or 16. That might be worth exploring. 

If `pycrc` needs a CRC algorithm that can be reflected in and out, I think you can take the complement of the in/out values. For example (not tested):

Field | Value
--- | ---
Width | 32
Poly | 0x04c11db7
XorIn | 0xefcbf201
ReflectIn | True
XorOut | 0x00000000
ReflectOut | True

On the topic of performance, I'm not sure which replays you used for your numbers. I typically test with [`3381.replay`](https://github.com/tfausak/rattletrap/blob/586196a704f83b975a9451bfdadeda6fb9c14d70/replays/3381.replay) because it's the biggest one in Rattletrap's test suite. It's 1.5 MB, which is pretty typical. The biggest replay I have from anywhere is [`a441.replay`](https://www.rocketleaguereplays.com/replays/a4416dea-40c8-17ef-51f3-4a8bb341942f/), which is 4.2 MB. 

There are obviously a lot of variables involved when benchmarking. I've been playing around with a minimal Haskell parser that parses everything *except* the frames. (So, you know, it avoids the hard part.) Using `3381.replay` from above, I routinely see it finish in around 10 ms. That's promising because it's about the same amount of time as computing the CRC (from the previous benchmark). 

Best case scenario, Rattletrap is only a little slower than Boxcars. There are two ways I could see Rattletrap getting the edge over Boxcars: Either using laziness to avoid parsing unwanted sections, or fanning the work out across CPU cores. However neither of those are very compelling. If the entire replay can be parsed in 20 ms, it doesn't make sense to save 10 ms by not parsing something. Similarly, the parser is going to be marching straight through a contiguous chunk of memory, so it's unlikely that the work can be productively split up. 

So, all that being said: Is there any reason for me to continue developing Rattletrap? As far as I know, nobody actually uses the Haskell library side of it. If the CLI can be replaced by Boxcars, there's not much reason for it to exist. On the other hand, now that Rattletrap isn't the fastest thing around, it could be fun to try and squeeze as much performance as possible out of it.

**Comment by nickbabcock** [comment]
> If pycrc needs a CRC algorithm that can be reflected in and out, I think you can take the complement of the in/out values. For example (not tested):

Ah it didn't work. [A PAINLESS GUIDE TO CRC ERROR DETECTION ALGORITHMS](http://www.repairfaq.org/filipg/LINK/F_crc_v34.html#CRCV_005) defines ReflectIn as:

> If it is FALSE, input bytes are processed with bit 7 being treated as the most significant bit (MSB) and bit 0 being treated as the least significant bit. If this parameter is FALSE, each byte is reflected before being processed.

This makes it sound like processing the data in a big endian order would fix things, but I haven't had such luck. Tried reversing the model using reveng. Oh well, like you mention, not a huge loss

> On the topic of performance, I'm not sure which replays you used for your numbers

D'oh, I forgot to mention 🤦‍♂️ , I was using [rumble.replay](https://github.com/nickbabcock/boxcars/blob/master/assets/rumble.replay) which is 1MB. I've appropriated 3381 and re-ran the benchmarks for it:

Parse header: 35us
Parse header + body: 18ms
Parse header + body + crc: 21ms
Parse header + body + crc + output json: 37ms

> So, all that being said: Is there any reason for me to continue developing Rattletrap?

Let me name a few:

- Rattletrap has more features (re-encoding a replay)
- Rattletrap parses a much larger collection than boxcars
- Rattletrap has improved error reporting for the network data (for now 😈 )
- Rattletrap is written in pure Haskell and is a good option for that community
- You now have leads on improving Rattletrap's performance -- and can try and flex your Haskell knowledge (and discover new tricks too)
- Interoperating Rust and Haskell is not flawless. [Curryrs](https://github.com/mgattozzi/curryrs) exists but definitely has limitations, so some [IPC](https://en.wikipedia.org/wiki/Inter-process_communication) might have to be used, which _may_ defeat some of the performance gain.
- The biggest reason: open source is not a competition. There's not a single winner and that's especially important when you're dealing with an uncontrolled input spec like rocket league. There are many ways to solve the same problem and it's not even about the solution at the end. It's doing something for yourself. Anyways, I'm getting too sentilmental

**Comment by tfausak** [comment]
Thanks for the kind words! Things aren't as grim as I made them out to be 😛 Honestly after a couple years of maintaining Octane/Rattletrap, I'm happy to have this kick in the pants to look at it with a critical eye again! Plus, it's not just you and me. We also have @jjbott doing all the hard work on his C# parser that I just copy from 😉 

I was curious how far Rattletrap has to go to catch up with Boxcars in terms of performance. Turns out: A lot. Like, a lot a lot. Two orders of magnitude. I added performance output to my test suite and saw these results: <https://gist.github.com/tfausak/cedf2b2df2063fb40ff9903507bc98cc>

```
3381: Decoded   1.484 MB in  2403.673 ms at  0.618 MB/s
```

That's the entire program execution time, so the numbers may be inflated a little. Even so, two and a half seconds is a far cry from 37 milliseconds. Lots of room for improvement. It should be fun to figure out how to make it faster!

**Comment by tfausak** [comment]
I've been working on a minimal parser written from scratch to explore the performance space with Haskell. I got it in good enough shape to parse `3381.replay`. It still has a ton of inefficiencies, but it can parse that replay in about 390ms. So it's only off by one order of magnitude. Seems doable!

**Comment by nickbabcock** [comment]
Awesome, are you pushing it to a branch? What optimizations have you used?

**Comment by tfausak** [comment]
It's completely separate at this point. I'm developing it in a Gist: <https://gist.github.com/tfausak/85e62a69129643f907ee4a6e581a3169>

It started off as an experiment to remove dependencies and kind of spiraled from there. The current bottleneck is getting the compressed 32-bit unsigned integers (Rattletrap calls them `CompressedWord`s, the script calls them `Max32`s). It's implemented in the most naive way right now, so it should be pretty easy to speed that part up. 

Avoiding string-based lookups is smart and easily doable. In fact, it might be possible to remove strings from the parser all together and use `ByteString`s (similar `Vec<u8>` in Rust) exclusively. 

There is an optimization that I think would be fun to try out: Parse each batch of keyframes independently. The list of keyframes gives you the frame number and bit offset for each keyframe. Since each keyframe replicates all actors, each chunk of frames starting with a keyframe can be considered in isolation. For example, with `3381.replay`:

``` yaml
num-frames: 13320
keyframes:
- number: 0
  time: 8.0
  frame: 0
  offset: 0
- number: 1
  time: 18.0
  frame: 286
  offset: 165702
- number: 2
  time: 28.0
  frame: 573
  offset: 432540
# ...
```

You could allocate an empty vector for the frames with a capacity of 13,320. Then you could parse frames 0 through 285 as a unit, 286 through 572, and so on. If you spread that out over *N* threads, it could potentially be up to *N* times faster since they're all independent.

**Comment by tfausak** [comment]
I should have done this benchmark a while ago: https://gist.github.com/tfausak/581e3bc51f7565d89f1798beb37469a5

It generates 2MB of random data and then walks through it bit-by-bit. This represents the absolute worst case performance scenario for parsing a replay. A lot of the information can be parsed in chunks, but sometimes you need to get one bit at a time. 

I am not encouraged by the results. The best case scenario for walking over 2MB was 297ms. That's pretty slow. It feels like the primitives available to me in Haskell aren't fast enough to compete with Rust. (I ran a similar benchmark in Rust and it completed in around 2ms.)

**Comment by nickbabcock** [comment]
```hs
-- TODO: This implementation is hideous.
makeClassAttributeMap :: List Cache ->
```
Comment much appreciated -- I was scrolling and that implementation stood out to me 🤣 

(I'm hypothesizing that) `ByteString` would be a small optimization. Parsing out strings has not shown on my profiling runs, but that may be because thhe string encoding library I use is [encoding_rs](https://github.com/hsivonen/encoding_rs), which doesn't need to allocate a string if the data is already in UTF-8 format, which most of the time it is.

Ah, I hadn't thought of using the keyframes to speed up parsing. That could work. On the other hand, an application could be parsing several replays at once (web server) and the concurrency may be better spent on parsing replays in parallel. 

> 2MB was 297ms

That's 7 bytes per microsecond, which does seem slow. Maybe there are other bit-based haskell projects that you can take inspiration from.

**Comment by tfausak** [comment]
That dang class attribute map is tricky! The implementation started off relatively simple. It got worse as I tried more replays. I honestly don't get why it's there at all; there's got to be a better way! 

The actual string decoding probably isn't a bottleneck. There aren't that many strings to be decoded in any given replay. I was thinking that sticking with `ByteString`s would encourage me to look up things by ID rather than by name. 

Also, for what it's worth, I *think* strings in replays are either Windows 1252 or UTF-16. I haven't completely confirmed that, but Rattletrap treats text as either Latin-1, which is similar to Windows 1252, or UTF-16. Of course the ASCII parts of UTF-8, Latin-1, and Windows 1252 are all the same.

Parallelizing frame parsing could definitely be more trouble than it's worth. As you point out, parsing multiple replays in parallel might be a better way to go. Also it's possible that splitting up the work would result in worse cache performance or something. It could be fun to try out anyway, maybe with [Rayon](https://crates.io/crates/rayon)? 

I wrote a pretty direct translation of my Haskell benchmark into Rust. It chewed through 2MB of bits in about 20ms, which comes to 108 bytes per microsecond. So for this synthetic benchmark, Rust is about 15x faster than Haskell. https://gist.github.com/tfausak/ec20235ffe2057dbcd14d84124f90ca7

To reiterate, this is a synthetic benchmark. When I benchmark what the core of Rattletrap actually looks like, it gets *a lot* slower. It takes 1.3s for the 2MB workload. That is only 1.6 bytes per microsecond. https://gist.github.com/tfausak/1882959b90742e9ecd53c28ee6942afb

I'm sure there are ways to speed up the Haskell benchmark to make it competitive with Rust. However I suspect that they will require writing code that looks like Rust, or even writing actual Rust or C and calling it through the FFI. 

Rust has other benefits too, like easily creating static binaries or cross compiling for unusual architectures like ARM or WASM. I think that it would be better for me to help make Boxcars feature compatible with Rattletrap than it would for me to try and make Rattletrap performance competitive with Boxcars. So, all that to say: How can I help? 😄

**Comment by nickbabcock** [comment]
> WASM

Wow, yeah I never thought about that. An in-browser, super speedy replay parser would be awesome

> How can I help?

I just created a bunch of issues, some of them should be beginner friendly like (https://github.com/nickbabcock/boxcars/issues/10 and https://github.com/nickbabcock/boxcars/issues/9). Not all of them relate to the network / body data, but most of them do.

I didn't create an issue for it, but expanding the ability to parse more replays and their network data. For me, this is labor intensive and involves a lot of `println` statements to pinpoint the issue 😄

**Comment by tfausak** [comment]
I left a few comments on some issues. I'll try to poke around in this repository and get a feel for it soon. I'm very new to Rust so I don't expect to be able to contribute right out of the gate. I'm happy to help out by giving English descriptions of how to parse something and/or translating my Haskell code into something more reasonable. 

Overall it seems like this issue can be closed :) I'm impressed that you were able to parse the network stream so quickly. Nice job!

**Comment by nickbabcock** [comment]
Boxcars is now able to parse all my replays -- there are still about 15 more from rattletrap, which I'll slowly chip away at 😄
